### PR TITLE
ZK-3913: Provide a barcode/QR code component

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zul/src/archive/metainfo/xml/zul.xsd [hunk 1]
```diff
@@ -956,6 +958,17 @@
...
+	<!--barcode-->
+	<xs:element name="barcode" type="barcodeType"/>
+	<xs:complexType name="barcodeType">
+		<xs:group ref="baseGroup" minOccurs="0" maxOccurs="unbounded"/>
+		<xs:attribute name="value" type="xs:string"/>
+		<xs:attribute name="type" type="xs:string" use="optional"/>
+		<xs:attribute name="height" type="xs:string" use="optional"/>
+		<xs:attribute name="displayValue" type="booleanType" use="optional"/>
+		<xs:attribute name="fontSize" type="intType" use="optional"/>
+		<xs:attribute name="_barWidth" type="intType" use="optional"/>
```
Comment (scribetw) [inline zul/src/archive/metainfo/xml/zul.xsd]: There is no attribute which is used `_` as suffix.

File: zul/src/archive/metainfo/xml/zul.xsd [hunk 2]
```diff
@@ -956,6 +958,17 @@
 		<xs:group ref="anyGroup" minOccurs="0" maxOccurs="unbounded" />
 		<xs:attributeGroup ref="xulElementAttrGroup" />
 	</xs:complexType>
+	<!--barcode-->
+	<xs:element name="barcode" type="barcodeType"/>
+	<xs:complexType name="barcodeType">
+		<xs:group ref="baseGroup" minOccurs="0" maxOccurs="unbounded"/>
+		<xs:attribute name="value" type="xs:string"/>
```
Comment (scribetw) [inline zul/src/archive/metainfo/xml/zul.xsd]: value can be omitted

File: zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		return TypeDate.getTypes();
+	}
+
+	public String getValue () {
+		return _value;
+	}
+	@NotifyChange({"value"})
+	public void setValue(String v) {
+		_value = v;
+		BindUtils.postNotifyChange(null, null, F85_ZK_3913VM.this, "value");
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java]: All your setters have `@NotifyChange`, why do you need to call `BindUtils`?

File: zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import java.util.List;
+
+public class F85_ZK_3913VM {
+//	String _value = "https://www.zkoss.org/";
+	String _value = "1234";
+	String _type = "CODE128";
+	String _height = "100px";
+	boolean _displayValue = false;
+	int _fontSize = 10;
+	int _barWidth = 2;
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java]: These fields should be private.

File: zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+import org.zkoss.bind.BindUtils;
+import org.zkoss.bind.annotation.AfterCompose;
+import org.zkoss.bind.annotation.Command;
+import org.zkoss.bind.annotation.NotifyChange;
+import org.zkoss.zk.ui.event.Event;
+import org.zkoss.zk.ui.select.SelectorComposer;
+import org.zkoss.zk.ui.select.annotation.Listen;
+import org.zkoss.zk.ui.select.annotation.Wire;
+import org.zkoss.zkmax.zul.Barcode;
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F85_ZK_3913VM.java]: Unused imports

File: zktest/src/archive/test2/F85-ZK-3913.zul [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+        <!--e = "";-->
+        <!--for (var ctx = canvas.getContext("2d"), y = 0, r = 0; r < barcode.num_rows; ++r) {-->
+        <!--for (var x = 0, c = 0; c < barcode.num_cols; ++c)-->
+        <!--1 == barcode.bcode[r][c] && ctx.fillRect(x, y, bw, bh), x += bw;-->
+        <!--y += bh-->
+        <!--};-->
+        <!--}, 'PDF417', ['pdf417']);-->
+        <!--});-->
+
+        <!--]]></script>-->
```
Comment (scribetw) [inline zktest/src/archive/test2/F85-ZK-3913.zul]: These code can be replaced with `<script src="./js/zk-3913-l.js" />`.

File: zktest/src/archive/test2/config.properties [hunk 7]
```diff
@@ -2852,6 +2852,8 @@ F85-ZK-3839.zul=A,E,ZUML,cache
 ##zats##F85-ZK-3816.zul=A,M,Proxy,ViewModel,NotifyChange,Auto
 F85-ZK-3808.zul=A,M,Listbox.Performance,ROD,BindClean
 ##ztl##F85-ZK-3835.zul=A,M,Tree,Client,ROD
+##zats##F85-ZK-1148.zul=A,M,annotation,@Destroy,mvvm,binder
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties]: When you rebased/merged, pay attention to unrelated code, okay? I mean it.

File: zktest/src/archive/WEB-INF/lang-addon.xml [hunk 8]
```diff
@@ -22,6 +22,9 @@ Copyright (C) 2006 Potix Corporation. All Rights Reserved.
 	<!-- you can specify addition JavaScript files
 	<javascript src="/js/my.js" charset="UTF-8"/>
 	-->
+    <javascript src="/test2/js/zk-3913-lib.js"/>
+    <javascript src="/test2/js/zk-3913-l.js"/>
+    <!-- assume you package it as /myjs/foo.js -->
```
Comment (scribetw) [inline zktest/src/archive/WEB-INF/lang-addon.xml]: why you added this comment???

File: zktest/src/archive/test2/F85-ZK-3913.zul [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+
+-->
+
+<zk>
+    <window viewModel="@id('vm')@init('org.zkoss.zktest.test2.F85_ZK_3913VM')">
+
+
+        <label multiline="true">
+            1. At the chosen box, you can change the type you like.
```
Comment (scribetw) [inline zktest/src/archive/test2/F85-ZK-3913.zul]: It is not a `chosenbox` but a `combobox`

File: zktest/src/archive/test2/config.properties [hunk 10]
```diff
@@ -2852,6 +2852,8 @@ F85-ZK-3839.zul=A,E,ZUML,cache
 ##zats##F85-ZK-3816.zul=A,M,Proxy,ViewModel,NotifyChange,Auto
 F85-ZK-3808.zul=A,M,Listbox.Performance,ROD,BindClean
 ##ztl##F85-ZK-3835.zul=A,M,Tree,Client,ROD
+##zats##F85-ZK-1148.zul=A,M,annotation,@Destroy,mvvm,binder
+##zats##F85-ZK-3913.zul=A,M,barcode
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties]: And where is your zats test case file `F85_ZK_3913Test.java`?

File: zktest/src/archive/test2/F86-ZK-3913.zul [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+
+-->
+
+<zk>
+    <window viewModel="@id('vm')@init('org.zkoss.zktest.test2.F86_ZK_3913VM')">
+
+
+        <label multiline="true">
+            1. At the combobox box, you can change the type you like.
```
Comment (scribetw) [inline zktest/src/archive/test2/F86-ZK-3913.zul]: just combobox

File: zktest/src/archive/test2/config.properties [hunk 12]
```diff
@@ -2854,6 +2854,7 @@ F85-ZK-3839.zul=A,E,ZUML,cache
 F85-ZK-3808.zul=A,M,Listbox.Performance,ROD,BindClean
 ##ztl##F85-ZK-3835.zul=A,M,Tree,Client,ROD
 F85-ZK-3875.zul=A,E,Textbox,submitByEnter
+##zats##F86-ZK-3913.zul=A,M,barcode
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties]: Put under `#F86`

File: zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+	public int getBarWidth() {
+		return _barWidth;
+	}
+
+	public void setBarWidth(int barWidth) {
+		_barWidth = barWidth;
+	}
+
+	@NotifyChange({"value"})
+	@Command("setValue123456789012")
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java]: If your command name is the same as method name, just `@Command`

File: zktest/src/archive/WEB-INF/lang-addon.xml [hunk 14]
```diff
@@ -22,6 +22,8 @@ Copyright (C) 2006 Potix Corporation. All Rights Reserved.
 	<!-- you can specify addition JavaScript files
 	<javascript src="/js/my.js" charset="UTF-8"/>
 	-->
+    <javascript src="/test2/js/zk-3913-PDF417lib.js"/>
```
Comment (scribetw) [inline zktest/src/archive/WEB-INF/lang-addon.xml]: This file uses Tab as indention, so follow that rule.

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+/* F85_ZK_3913Test.java
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java]: F86

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+public class F85_ZK_3913Test extends WebDriverTestCase {
+
+	@Test
+	public void test() {
+		connect();
+		JQuery canvas = jq("canvas");
+		assertEquals(3, canvas.length());
+		
+		String height = getEval("jq(\"canvas\")[0].height");
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java]: `jq("canvas").eq(0).height()`

`getEval` is the last resort while this is no such method to be used.

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+		waitResponse();
+		String width = getEval("jq(\"canvas\")[0].width");
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java]: `jq("canvas").eq(0).width()`

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+		waitResponse();
+		assertEquals("100", height);
+		assertEquals("114", width);
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java]: I don't think checking the height/width is a good idea.

Checking if something being drawn may be better: https://stackoverflow.com/q/4644524

File: zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+	@Command("changeBarWidth")
+	public void changeBarWidth() {
+		if (_barWidth == 2) {
+			setBarWidth(1);
+		} else {
+			setBarWidth(2);
+		}
+	}
+
+	public static class Type {
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java]: Why not just
```java
private static final List<String> TYPES = Collections.unmodifiableList(
    Arrays.asList("CODE39", ...)
);
```

File: zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+                
+        History:
+                Tue Apr 24 4:22 PM:55 CST 2018, Created by klyve
+
+Copyright (C) 2018 Potix Corporation. All Rights Reserved.
+*/
+package org.zkoss.zktest.zats.test2;
+
+import org.junit.Test;
+import org.zkoss.zats.mimic.ComponentAgent;
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/F85_ZK_3913Test.java]: Please check your unused import before commiting

File: zktest/src/archive/test2/F86-ZK-3913.zul [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+            1. At the combobox, you can change the type you like.
+            2. Click the button under "Value:" you want to choose, you can render the barcode as the value you choose.
+            3. Click the button under "Change_Other_Attribute", you can change the barcode attribute.
+        </label>
+        <div>
+            value:
+            <label value="@bind(vm.value)"/>
+        </div>
+        <div height="100px">
+            <barcode id="brcode" type="@bind(vm.type)" value="@bind(vm.value)" height="@bind(vm.height)"
```
Comment (scribetw) [inline zktest/src/archive/test2/F86-ZK-3913.zul]: For read only use `@load` instead of `@bind`

File: zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.List;
+
+public class F86_ZK_3913VM {
+	private String _value = "1234";
+	private String _type = "CODE128";
+	private String _height = "100px";
+	private boolean _displayValue = false;
+	private int _fontSize = 10;
+	private int _barWidth = 2;
+	List<String> types = Collections.unmodifiableList(Arrays.asList("CODE39", "CODE128", "CODE128A", "CODE128B",
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java]: As I previous said,`private static final List<String> TYPES` is better.

File: zktest/src/archive/test2/F86-ZK-3913.zul [hunk 23]
```diff
@@ -0,0 +1,75 @@
...
+
+
+        <label multiline="true">
+            1. At the combobox, you can change the type you like.
+            2. Click the button under "Value:" you want to choose, you can render the barcode as the value you choose.
+            3. Click the button under "Change_Other_Attribute", you can change the barcode attribute.
+        </label>
+        <div>
+            value:
+            <label value="@bind(vm.value)"/>
```
Comment (scribetw) [inline zktest/src/archive/test2/F86-ZK-3913.zul]: For read only use `@load` instead of `@bind`

File: zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java [hunk 24]
```diff
@@ -0,0 +1,148 @@
...
+import java.util.List;
+
+public class F86_ZK_3913VM {
+	private String _value = "1234";
+	private String _type = "CODE128";
+	private String _height = "100px";
+	private boolean _displayValue = false;
+	private int _fontSize = 10;
+	private int _barWidth = 2;
+	private static final List<String> types = Collections.unmodifiableList(Arrays.asList("CODE39", "CODE128", "CODE128A", "CODE128B",
```
Comment (scribetw) [inline zktest/src/org/zkoss/zktest/test2/F86_ZK_3913VM.java]: TYPES

**Comment by wenninghsu** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I could not rebase F85-ZK-3913 with 8.6. There are conflicts.
